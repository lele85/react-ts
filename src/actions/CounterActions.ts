import actionCreatorFactory from 'typescript-fsa';

const actionCreator = actionCreatorFactory("@@counter");
export const counterInc = actionCreator<{by: number}>('INC');
export const counterDec = actionCreator<{by: number}>('DEC');