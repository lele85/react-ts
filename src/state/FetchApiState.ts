export enum ApiActionStatus {
    LOADING="LOADING",
    SUCCESS="SUCCESS",
    ERROR="ERROR"
};

export type FetchApiState<Tmodel> = {
    status: ApiActionStatus,
    model : Tmodel | null
};