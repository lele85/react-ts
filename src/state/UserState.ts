export interface UserState {
    readonly firstname: string;
    readonly lastname: string;
}
