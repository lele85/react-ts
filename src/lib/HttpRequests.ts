import { set, get } from "lodash";
import { AbortTokens } from "../actions/ShowActions";
import Axios from "axios";

const IDENTITY = (x: any) => x;

const requests = {};

const getHttpRequest = (token: AbortTokens) => {
    return get(requests, `${token}`, { cancel: IDENTITY, token: null });
};

const setHttpRequest = (token: AbortTokens) => {
    if (token) {
        const source = Axios.CancelToken.source();
        set(requests, `${token}`, source);
    }
};

const abortHttpRequest = (token: AbortTokens) => {
    getHttpRequest(token).cancel();
};

export default {
    getHttpRequest,
    setHttpRequest,
    abortHttpRequest
};
