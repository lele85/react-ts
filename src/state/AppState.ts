import { CounterState } from "./CounterState";
import { UserState } from "./UserState";
import { FetchApiState } from "./FetchApiState";
import { IShow } from "../model/IShow";

export type AppState = {
    readonly counter: CounterState,
    readonly user : UserState,
    readonly shows: FetchApiState<IShow[]>
};