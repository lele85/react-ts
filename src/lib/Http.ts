import Axios from "axios";
import { isNullOrUndefined } from "util";
import { AbortTokens } from "../actions/ShowActions";
import HttpRequests from "./HttpRequests";

export type FetchParams = {
    [key: string]: string | number;
};

export type HttpResponse<T> = { canceled: boolean; model: T | null };

export function buildUri(url: string, params: FetchParams) {
    let remainingParams = {
        ...params
    };

    const tokens = url.split("/").map(token => {
        if (!token.startsWith(":")) {
            return token;
        }
        const key = token.substr(1);
        if (!isNullOrUndefined(params[key])) {
            delete remainingParams[key];
            return params[key];
        }
        return token;
    });

    return {
        uri: tokens.join("/"),
        params: remainingParams
    };
}

function get<TResponse, TParams extends FetchParams>(
    uriPlaceholder: string,
    allParams: TParams,
    abortToken: AbortTokens
): Promise<HttpResponse<TResponse>> {
    const { uri, params } = buildUri(uriPlaceholder, allParams);

    HttpRequests.abortHttpRequest(abortToken);
    HttpRequests.setHttpRequest(abortToken);
    const request = HttpRequests.getHttpRequest(abortToken);

    return Axios.get(uri, { ...params, cancelToken: request.token })
        .then(res => {
            const response: TResponse = res.data;
            return {
                canceled: false,
                model: response
            };
        })
        .catch(e => {
            if (Axios.isCancel(e)) {
                return Promise.resolve({
                    canceled: true,
                    model: null
                });
            } else {
                return Promise.reject(e);
            }
        });
}

export default {
    get: get
};
