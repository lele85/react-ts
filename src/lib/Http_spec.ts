import Http, { buildUri } from "./Http";
import { AbortTokens } from "../actions/ShowActions";
import Axios from "axios";
import HttpRequests from "./HttpRequests";
describe("Http", () => {
    describe("get", () => {
        const mockModel = {
            id: 12,
            title: "Show Title"
        };
        const mockParams = {
            id: 12,
            exports: "all"
        };
        const mockCancel = {
            cancel: () => {},
            token: "123"
        };

        beforeEach(() => {
            jest.spyOn(HttpRequests, "getHttpRequest").mockReturnValue(mockCancel);
            jest.spyOn(HttpRequests, "setHttpRequest");
            jest.spyOn(HttpRequests, "abortHttpRequest");
        });

        it("should cancel the pending request with the same token, set the current one and resolve succesfully", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.resolve({ data: mockModel }));
            const result = await Http.get<typeof mockModel, typeof mockParams>(
                "/api/shows/:id",
                mockParams,
                AbortTokens.SHOW
            );
            expect(result).toEqual({ canceled: false, model: mockModel });
            expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
        });

        it("should cancel the pending request with the same token, set the current one and reject in case of error", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.reject({ error: new Error("message") }));
            try {
                await Http.get<typeof mockModel, typeof mockParams>("/api/shows/:id", mockParams, AbortTokens.SHOW);
            } catch (e) {
                expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
            }
        });

        it("should cancel the pending request with the same token, set the current one and resolve with a canceled response in case of CancelationError", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.reject(new Axios.Cancel("message")));
            const result = await Http.get<typeof mockModel, typeof mockParams>(
                "/api/shows/:id",
                mockParams,
                AbortTokens.SHOW
            );
            expect(result).toEqual({ canceled: true, model: null });
            expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
        });
    });

    describe("buildUri", () => {
        it("should return the same uri if no placeholder and no params", () => {
            expect(buildUri("/foo/bar", {})).toEqual({ uri: "/foo/bar", params: {} });
        });

        it("should replace the placholder if there's a match and remove the bound param", () => {
            expect(buildUri("/foo/bar/:id", { id: 12 })).toEqual({
                uri: "/foo/bar/12",
                params: {}
            });
        });

        it("should replace the placholder if there's a match and remove the bound param", () => {
            expect(buildUri("/foo/bar/:id", { id: 12, q: "search" })).toEqual({
                uri: "/foo/bar/12",
                params: { q: "search" }
            });
        });

        it("should replace the placeholedet with a match for value 0", () => {
            expect(buildUri("/foo/bar/:id", { id: 0, q: "search" })).toEqual({
                uri: "/foo/bar/0",
                params: { q: "search" }
            });
        });
    });
});
