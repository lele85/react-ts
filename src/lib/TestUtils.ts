import Http from "./Http";
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

export function mockGetSuccess<T>(result : T) {
    Http.get = jest.fn<T>(() => {
        return Promise.resolve(result);
    });
};

export function mockGetError<T>(error : T) {
    Http.get = jest.fn<T>(() => {
        return Promise.reject(error);
    });
};

export function createMockStore<T>(state: T) {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore(state);
}

export function assertEquals<T> (result : T, expected: T) {
    return expect(result).toEqual(expected);
}