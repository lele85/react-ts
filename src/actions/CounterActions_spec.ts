import { CounterState } from '../state/CounterState';
import { createMockStore } from '../lib/TestUtils';
import { counterInc, counterDec } from './CounterActions';

describe("Counter Actions", () => {

    it("should be dispatched", async () => { 
        const store = createMockStore<CounterState>({value:0});
        store.dispatch(counterInc({by:2}));
        store.dispatch(counterDec({by:3}));
        store.dispatch(counterInc({by:10}))
        expect(store.getActions()).toMatchSnapshot();
    });

});