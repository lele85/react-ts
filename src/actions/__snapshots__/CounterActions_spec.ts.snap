// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Actions should be dispatched 1`] = `
Array [
  Object {
    "payload": Object {
      "by": 2,
    },
    "type": "@@counter/INC",
  },
  Object {
    "payload": Object {
      "by": 3,
    },
    "type": "@@counter/DEC",
  },
  Object {
    "payload": Object {
      "by": 10,
    },
    "type": "@@counter/INC",
  },
]
`;
